File: common/src/main/java/dev/kineticcat/hexportation/api/transfer/Storage.java
package dev.kineticcat.hexportation.api.transfer;

import java.util.Iterator;

/**
 * Cross-platform replacement for Fabric's Storage interface.
 * Provides identical API so existing code needs zero changes.
 */
public interface Storage<T> {
    
    /**
     * Insert into this storage.
     */
    long insert(T resource, long maxAmount, Object transaction);
    
    /**
     * Simulate insertion into this storage.
     */
    long simulateInsert(T resource, long maxAmount, Object transaction);
    
    /**
     * Get an iterator over all non-empty storage views.
     */
    Iterator<StorageView<T>> nonEmptyIterator();
    
    /**
     * Get an iterator over all storage views.
     */
    Iterator<StorageView<T>> iterator();
    
    /**
     * Extract from this storage.
     */
    long extract(T resource, long maxAmount, Object transaction);
    
    /**
     * Simulate extraction from this storage.
     */
    long simulateExtract(T resource, long maxAmount, Object transaction);
}

File: common/src/main/java/dev/kineticcat/hexportation/api/transfer/Storage.kt
package dev.kineticcat.hexportation.api.transfer

/**
 * Cross-platform replacement for Fabric's Storage interface.
 * Provides identical API so existing code needs zero changes.
 */
interface Storage<T> {
    
    /**
     * Insert into this storage.
     */
    fun insert(resource: T, maxAmount: Long, transaction: Any?): Long
    
    /**
     * Simulate insertion into this storage.
     */
    fun simulateInsert(resource: T, maxAmount: Long, transaction: Any?): Long
    
    /**
     * Get an iterator over all non-empty storage views.
     */
    fun nonEmptyIterator(): Iterator<StorageView<T>>
    
    /**
     * Get an iterator over all storage views.
     */
    fun iterator(): Iterator<StorageView<T>>
    
    /**
     * Extract from this storage.
     */
    fun extract(resource: T, maxAmount: Long, transaction: Any?): Long
    
    /**
     * Simulate extraction from this storage.
     */
    fun simulateExtract(resource: T, maxAmount: Long, transaction: Any?): Long
}